llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1250,6 +1265,158 @@ void AddressSanitizerPass::printPipeline(
   OS << '>';
 }
 
+// Append a new argument "launch_data" to user's spir_kernel & spir_func
+static void ExtendSpirKernelArgs(Module &M, FunctionAnalysisManager &FAM) {
+  SmallVector<Function *> SpirFixupFuncs;
+  for (Function &F : M) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
+      SpirFixupFuncs.emplace_back(&F);
+    } else if (F.getCallingConv() == CallingConv::SPIR_FUNC) {
+      if (M.getModuleFlag("openmp-device")) {
+        const TargetLibraryInfo &TLI = FAM.getResult<TargetLibraryAnalysis>(F);
+        LibFunc LF;
+        if (TLI.getLibFunc(F, LF)) {
+          continue;
+        } else {
+          if (F.getName().starts_with("_Z") &&
+              F.getName().find("__spirv_") != StringRef::npos)
+            continue;
+
+          if (F.getName().equals("_Z12get_local_idj") ||
+              F.getName().equals("_Z14get_local_sizej") ||
+              F.getName().equals("_Z14_get_num_groupsj"))
+            continue;
+        }
+
+      } else {
+        if (F.isDeclaration() || F.getName().starts_with("__sycl_") ||
+            F.getName().starts_with("__spirv") ||
+            F.getFunctionType()->isVarArg())
+          continue;
+      }
+      SpirFixupFuncs.emplace_back(&F);
+    }
+  }
+
+  SmallVector<std::pair<Function *, Function *>> SpirFuncs;
+  const int LongSize = M.getDataLayout().getPointerSizeInBits();
+  auto *IntptrTy = Type::getIntNTy(M.getContext(), LongSize);
+
+  for (auto *F : SpirFixupFuncs) {
+    SmallVector<Type *, 16> Types;
+    for (Function::const_arg_iterator I = F->arg_begin(), E = F->arg_end();
+         I != E; ++I) {
+      Types.push_back(I->getType());
+    }
+
+    // New Argument
+    Types.push_back(IntptrTy);
```

Comment:
is it right that the new arg points to a global memory? If so, why are we using integer type for it.

Reply:
I use integer type for simplicity. In fact, it's type should be `__SYCL_GLOBAL__ LaunchInfo*`, or simply `__SYCL_GLOBAL__ void*`.

Reply:
I think it is probably better to use its real type:
1. metadata in `!kernel_arg_addr_space !kernel_arg_access_qual !kernel_arg_type !kernel_arg_base_type !kernel_arg_type_qual` will be correct.
2. the backend compiler is aware that it is a global buffer.
3. easy for people to understand the IR.

Reply:
Ok, I'll change its type to "i64 as(1)*".